python/whylogs/api/logger/experimental/logger/actor/actor.py

```diff
@@ -0,0 +1,232 @@
+    Configuration for the queue used by the actor.
+
+    Parameters
+    ----------
+    max_batch_size : int
+        Maximum number of messages to process in a single batch.
+    message_accumualtion_duration : float
+        How long to accumulate messages before processing a batch
+    message_poll_wait : float
+        How long to block when polling for messages in a loop.
+    """
+
+    max_batch_size: int = 50_000
+    message_accumualtion_duration: float = 1.0  # seconds
+    message_poll_wait: float = 0.1  # seconds
+
+
+MessageType = TypeVar("MessageType")
+Messages = Union[MessageType, CloseMessage]
+
+
+class Actor(ABC, Generic[MessageType]):
+    """
+    Simple actor inspired interface designed for batch processing messages.
+    This was created with thread and process based whylogs logging in mind and it has
+    all of the common message passing logic needed.
+    """
+
+    def __init__(
+        self,
+        queue_wrapper: QueueWrapper[Messages],
+        queue_config: QueueConfig = QueueConfig(),
+    ) -> None:
+        self._queue: QueueWrapper[Messages] = queue_wrapper
+        self._logger = logging.getLogger(f"ai.whylabs.actor.{type(self).__name__}")
+        self._logger.setLevel(logging.DEBUG)
+        self._queue_config = queue_config
+        super().__init__()
+
+    @abstractmethod
+    def close_message_handled(self) -> bool:
+        raise NotImplementedError()
+
+    @abstractmethod
+    def set_close_message_handled(self) -> None:
+        raise NotImplementedError()
+
+    @abstractmethod
+    def close_message_wait(self) -> None:
+        raise NotImplementedError()
+
+    @abstractmethod
+    def is_done(self) -> bool:
+        raise NotImplementedError()
+
+    @abstractmethod
+    def done_wait(self) -> None:
+        raise NotImplementedError()
+
+    @abstractmethod
+    def set_done(self) -> None:
+        """
+        Set this actor as done, meaning it has finished processing all messages.
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def is_closed(self) -> bool:
+        raise NotImplementedError()
+
+    @abstractmethod
+    def set_closed(self) -> None:
+        """
+        Sets this actor as closed, meaning it should no longer accept messages.
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def process_batch(self, batch: List[Messages], batch_type: Type) -> None:
+        raise NotImplementedError()
+
+    def send(self, message: Messages) -> None:
+        if self.is_closed():
+            raise Exception("Actor is closed, can't send message.")
+
+        if isinstance(message, CloseMessage):
+            self.set_closed()
+
+        done = False
+        while not done:
+            try:
+                self._queue.send(message, timeout=self._queue_config.message_poll_wait)
+                done = True
+            except queue.Full:
+                current_size = self._queue.size()
+                if current_size == 0:
```

Comment:
If the queue is full, I would expect that the size() would return something above 0, is this code correct?

Reply:
The else branch is doing what you expect. This case only happens when the message is both full and effectively empty, which means it was configured with a max buffer size that is smaller than the messages that its trying to receive. This is how that shows up, at least in the faster-fifo lib. I've never actually encountered this outside of testing for it though. I added a bit more to the message to reflect that its hopeless.